lib/posix-socket/driver.c

```diff
@@ -0,0 +1,154 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR SOCKETINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <inttypes.h>
+#include <uk/alloc.h>
+#include <uk/assert.h>
+#include <uk/print.h>
+#include <uk/list.h>
+#include <uk/socket_driver.h>
+#include <sys/types.h>
+
+/**
+ * List of socket family drivers
+ */
+UK_TAILQ_HEAD(posix_socket_driver_list, struct posix_socket_driver);
+struct posix_socket_driver_list posix_socket_families =
+		UK_TAILQ_HEAD_INITIALIZER(posix_socket_families);
+static uint16_t uk_posix_socket_family_count;
+
+unsigned int
+posix_socket_family_count(void)
+{
+	return uk_posix_socket_family_count;
+}
+
+void _posix_socket_family_register(struct posix_socket_driver *d,
+		int fam,
+		struct posix_socket_ops *ops,
+		struct uk_alloc *alloc)
+{
+	UK_ASSERT(d);
+	UK_ASSERT(ops);
+
+	if (unlikely(fam == 0)) {
+		uk_pr_crit("Cannot register unspecified socket family: %p\n", ops);
+		return;
+	}
+
+	uk_pr_debug("Registering socket handler: library: %s, family: %d: %p...\n", d->libname, fam, ops);
+
+	struct posix_socket_driver *e;
+	UK_TAILQ_FOREACH(e, &posix_socket_families, _list) {
+		if (e->af_family == fam) {
+			uk_pr_warn("Socket family (%d) already registerd to library %s!", fam, e->libname);
+			return;
+		}
+	}
+
+	new_posix_socket_family(d, fam, alloc, ops);
+	UK_TAILQ_INSERT_TAIL(&posix_socket_families, d, _list);
+	uk_posix_socket_family_count++;
+}
+
+void
+_posix_socket_family_unregister(struct posix_socket_driver *d)
+{
+	UK_ASSERT(d);
+	UK_ASSERT(uk_posix_socket_family_count > 0);
+
+	uk_pr_debug("Unregistering socket handler: %d\n", d->af_family);
+
+	UK_TAILQ_REMOVE(&posix_socket_families, d, _list);
+	uk_posix_socket_family_count--;
+}
+
+static int
+posix_socket_family_init(struct posix_socket_driver *d)
+{
+	UK_ASSERT(d);
+	UK_ASSERT(d->ops);
+
+	uk_pr_debug("Initializing socket handler: %d\n", d->af_family);
+
+	/* Assign the default allocator if not specified */
+	if (d->allocator == NULL) {
+		d->allocator = uk_alloc_get_default();
+		uk_pr_debug("Using default memory allocator for socket family: %d: %p\n", d->af_family, d->allocator);
+	}
+
+	if (!d->ops->init)
+		return 0;
+
+	return d->ops->init(d);
+}
+
+/* Returns the number of successfully initialized device sockets */
+static int
+posix_socket_family_lib_init(void)
+{
+	uk_pr_info("Initializing socket handlers...\n");
```

Comment:
Just a nitpick: I would move this message after the variable declarations. Again, it's just a nitpick, in the worst case scenario it will cause a warning.